.github/workflows/tests.yml

```
@@ -477,30 +343,102 @@ jobs:
+      test-dependents: ${{ steps.check-labels.outputs.test-dependents }}
+      timeout-minutes: ${{ steps.check-labels.outputs.timeout-minutes }}
+      test-bot-formulae-args: ${{ steps.check-labels.outputs.test-bot-formulae-args }}
+      test-bot-dependents-args: ${{ steps.check-labels.outputs.test-bot-dependents-args }}
+      cleanup-linux-runner: ${{ steps.check-labels.outputs.cleanup-linux-runner }}
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Check for CI labels
+        id: check-labels
+        uses: actions/github-script@v6
+        env:
+          TESTING_FORMULAE: ${{needs.formulae_detect.outputs.testing_formulae}}
+          ADDED_FORMULAE: ${{needs.formulae_detect.outputs.added_formulae}}
+          DELETED_FORMULAE: ${{needs.formulae_detect.outputs.deleted_formulae}}
+          SCRIPTS_PATH: .github/scripts
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          retries: 3
+          script: |
+            const path = require('path')
+            const script = require(path.resolve(`${process.env.SCRIPTS_PATH}/check-labels.js`))
+
+            const formulae_detect = {
+              testing_formulae: `${process.env.TESTING_FORMULAE}`,
+              added_formulae: `${process.env.ADDED_FORMULAE}`,
+              deleted_formulae: `${process.env.DELETED_FORMULAE}`
+            }
+
+            try {
+              await script({github, context, core}, formulae_detect)
+            } catch (error) {
+              console.error(error);
+            }
+
+  setup_dep_runners:
+    needs: [formulae_detect, setup_dep_tests]
+    if: >
+      github.event_name == 'pull_request' &&
+      !fromJson(needs.setup_dep_tests.outputs.syntax-only) &&
+      fromJson(needs.setup_dep_tests.outputs.test-dependents)
+    runs-on: ubuntu-22.04
+    container:
+      image: ghcr.io/homebrew/ubuntu22.04:master
+    outputs:
+      runners: ${{steps.determine-dependent-runners.outputs.runners}}
+      runners_present: ${{steps.determine-dependent-runners.outputs.runners_present}}
+    env:
+      HOMEBREW_LINUX_RUNNER: ${{needs.setup_dep_tests.outputs.linux-runner}}
+      HOMEBREW_LINUX_CLEANUP: ${{needs.setup_dep_tests.outputs.cleanup-linux-runner}}
+      TESTING_FORMULAE: ${{needs.formulae_detect.outputs.testing_formulae}}
+    steps:
+      - name: Set up Homebrew
+        id: set-up-homebrew
+        uses: Homebrew/actions/setup-homebrew@master
+
+      - name: Determine runners to use for test_deps job
+        id: determine-dependent-runners
+        run: brew determine-test-runners --dependents --eval-all "$TESTING_FORMULAE"
+
   test_deps:
-    needs: [tap_syntax, setup_tests, setup_dep_runners, tests]
+    needs: [tap_syntax, setup_dep_tests, setup_dep_runners, tests]
     if: >
       (success() || 
-      (failure() && contains(github.event.pull_request.labels.*.name, 'CI-no-fail-fast'))) &&
+      (failure() && !fromJson(needs.setup_dep_tests.outputs.fail-fast))) &&
```

Probably need to make sure `tests` is not skipped. Otherwise: https://github.com/Homebrew/homebrew-core/actions/runs/4779050467

```suggestion
       (failure() &&
        !fromJson(needs.setup_dep_tests.outputs.fail-fast) &&
        needs.tests.result != 'skipped')) &&
```

---

Should that be
```suggestion
       (failure() &&
        !fromJson(needs.setup_dep_tests.outputs.fail-fast) &&
        needs.tests.result == 'failure')) &&
```
?

---

Or, maybe even better:

```suggestion
      (failure() &&
       !fromJson(needs.setup_dep_tests.outputs.fail-fast) &&
       needs.tap_syntax.result == 'success')) &&
```

---

I'm not sure if that does what's intended -- IIUC a `tap_syntax` success doesn't imply `tests`'s status (`tests` can still be skipped). Or if you mean `tests` here -- do we not want deps to be tested when `tests` fails? IMO this is one of the reasons why we have `CI-no-fail-fast`.

---

In the example run you give above `tests` was `skipped` because `tap_syntax` failed, so it would at least handle the example you've shown.

The problem with doing `needs.tests.result != 'skipped'` is that we can have `needs.tests.result == 'cancelled'`, but I don't think we want to run `test_deps` for that either.

---

Ok, that makes sense!